### DISCUSSION TITLE
Canceling a search from the grep crate

### DISCUSSION THREAD

**Discussion by chipsenkbeil** [BurntSushi/ripgrep#2294] [General] [answered]
I'm looking into adding search capabilities to https://github.com/chipsenkbeil/distant and am investigating using the crates that power ripgrep to do this for the reference implementation.

A feature I'd like to support is the ability to cancel an ongoing search since the files I'm querying could be large.

Is there a way to do this in the current crate implementations? I know I can write logic to cancel in between file searches, but wasn't sure if there was a way to cancel a searcher itself.

I'm going to spawn this in a tokio blocking task, so being able to still cancel would be helpful.

**Comment by BurntSushi** [comment, marked as answer]
The only way is to send a signal somehow, but I honestly don't know how that would work in practice.

Whether something more deliberate could be added is not entirely clear. Certainly, it simply cannot be done in all cases. For example, a multi-line search requires that the entire file be in memory (whether via memory map or heap), and then let's the regex engine run on it. There is no way, and probably never will be, a way to stop the regex engine once a search has started.

However, a line oriented search could in theory be stopped because it can search the file in segments via buffering. You couldn't interrupt a search on a single buffer, but you could, in theory, implement something to prevent search from buffering more contents and thus quitting the search there.

However, my experience tells me that adding cancellation to something that hasn't been designed for it initially is difficult, to say the least. The refactor to achieve it probably quite gnarly, particularly given that it cannot be done in all cases unless multi-line search is banned.

So in summary, I would say this is unlikely to be possible.

**Reply by BurntSushi** [reply]
It is worth noting that the grep-searcher crate can search an io::Read, which means you can in theory return EOF whenever you want. That is an avenue to explore that might work today. (But again, only for line oriented searches. A multi-line search will try to read everything from the io::Read up front before actually starting the search.)

IMO, your best bet is to put the search in a sub-process and just kill that process if you want to cancel the search.

**Reply by BurntSushi** [reply]
The sub-process approach would handle all cases, including multi-line search.

**Reply by chipsenkbeil** [reply]
Hadn't thought of the subprocess approach. I'll give that some thought. Thanks!

**Reply by chipsenkbeil** [reply]
@BurntSushi just want to give some quick thanks for the work you've done on this and many other Rust projects. I've just about gotten the `grep` crate integrated into my server to supply the searching logic. It's going to be a big help for remotely querying files from neovim!

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/2481802/187313828-a5a8cabb-c4f9-4a16-9b83-ec9d2a9a6510.png">

**Reply by BurntSushi** [reply]
w00t! Great to hear it!
